### DISCUSSION TITLE
Refactor the vehicle class

### DISCUSSION THREAD

**Discussion by frutiemax** [OpenRCT2/OpenRCT2#18918] [Feature requests]
The vehicle class code needs some refactoring:

- Some methods have nothing to do with each others, for example, there is a member method UpdateTrackMotionMiniGolfVehicle and there is one for UpdateSceneryDoor. These two methods should not be part of the same class.
- The vehicle update function checks for every state, even those that are not possible with the current operating mode. For example, in the Update function, it checks for the Vehicle::Status::TravellingDodgems and in the same switch case for Vehicle::Status::TopSpinOperating; those shouldn't be checked at the same time as the operating mode should not permit those two status.

I'd like to see those two above issues solved as it would make it easier to expose the vehicle update code in a way like the PaintFunctions in the RideTypeDescriptors. I had in mind to remove all those Update functions in the vehicle class and have functions that modify the vehicle object that depend on its current operating mode, i.e. GetUpdateFunction(OperatingMode mode).

**Comment by spacek531** [comment]
Why should UpdateTrackMotionMiniGolfVehicle not be part of Vehicle?

**Reply by frutiemax** [reply]
It's because you have too many unrelated functions in the same class, MiniGolf, SpaceRings, BoatLocations, etc. The vehicle class is currently doing too many things, it needs to be split up into smaller objects.

**Reply by spacek531** [reply]
Where are you going to put BoatLocations if it isn't in a Vehicle class?

**Reply by frutiemax** [reply]
> Where are you going to put BoatLocations if it isn't in a Vehicle class?

It seems like it belongs to a ride type? It would make sense to put this in the RideTypeDescriptor with an additional field.

**Reply by spacek531** [reply]
How would you put this in the RTD look at the references for BoatLocation

**Reply by frutiemax** [reply]
It's something I haven't thought about... Maybe the operating mode should actually be a class and contains those Update functions and related variables i.e. OperatingMode< RideMode > to avoid inheritance. This way, the OperatingMode< BoatHire > would be the only operating mode to contains variables related to boats. OperatingMode would be contained by the vehicle struct so it can be read from and written to when saving and loading parks.

**Reply by spacek531** [reply]
Having the update functions be in a separate class I could see working

**Reply by frutiemax** [reply]
I had a PR that did a similar thing but I was told not to use inheritance as it introduces calls to virtual tables... I might reopen it as it seemed a very logical way to do this.

**Comment by spacek531** [comment]
Are you planning on making extensions of a base vehicle class? How will you differentiate the types?

**Reply by frutiemax** [reply]
I would create functions that modify the vehicle structure, not make child classes. The update functions would be of signature void(*)(Vehicle& vehicle). Those functions would be in a table and we get the correct one with GetUpdateFunction(RideMode mode). Those functions would be friend of the vehicle class, or just make the vehicle fields all public since it shouldn't be a class but just a data type...

**Comment by duncanspumpkin** [comment]
It's very hard to untangle a lot of this without breaking existing parks. It would be good to start breaking up the vehicle into multiple components and yes perhaps reverting some member functions back to free functions.
